src/coreclr/gc/gc.cpp

```diff
@@ -11259,6 +11323,51 @@ void gc_heap::set_region_gen_num (heap_segment* region, int gen_num)
     assert (gen_num < (1 << (sizeof (uint8_t) * 8)));
     assert (gen_num >= 0);
     heap_segment_gen_num (region) = (uint8_t)gen_num;
+
+    uint8_t* region_start = get_region_start (region);
+    uint8_t* region_end = heap_segment_reserved (region);
+
+    size_t region_index_start = get_basic_region_index_for_address (region_start);
+    size_t region_index_end = get_basic_region_index_for_address (region_end);
+    uint8_t entry = (uint8_t)((gen_num << 4) | gen_num);
+    for (size_t region_index = region_index_start; region_index < region_index_end; region_index++)
+    {
+        assert (gen_num <= max_generation);
+        map_region_to_generation[region_index] = entry;
+    }
+    if (gen_num <= soh_gen1)
+    {
+        bool success_low = false;
+        bool success_high = false;
+        bool ephemeral_change = false;
+        while (!(success_low && success_high))
+        {
+            uint8_t* current_ephemeral_low = ephemeral_low;
+            if (current_ephemeral_low <= region_start)
+                success_low = true;
+            else
+            {
+                success_low = (Interlocked::CompareExchangePointer (&ephemeral_low, region_start, current_ephemeral_low) == current_ephemeral_low);
+                if (success_low)
+                    ephemeral_change = true;
+            }
+
+            uint8_t* current_ephemeral_high = ephemeral_high;
+            if (current_ephemeral_high >= region_end)
+                success_high = true;
+            else
+            {
+                success_high = (Interlocked::CompareExchangePointer (&ephemeral_high, region_end, current_ephemeral_high) == current_ephemeral_high);
+                if (success_high)
+                    ephemeral_change = true;
+            }
+        }
+        if (ephemeral_change)
+        {
+            stomp_write_barrier_ephemeral (ephemeral_low, ephemeral_high,
```

Comment:
stomp_write_barrier_ephemeral is required to be called while the EE is suspended. if we are calling this from init_heap_segment, it means it can be called when a new gen0 region is acquired while the EE is running.

Reply:
and when it's called from find_first_valid_region, we could optimize and only call this once at the end of the GC (if the ephemeral range actually got larger).